### ISSUE TITLE
Question on XPath predicates

### ISSUE THREAD

**Issue by zme1** [obdurodon/dh_course#208]
So I am stuck on question #3 on the homework.

My main problem is how to write an xpath that differentiates the types of Div. In other words, how am I suppose to write the xpath that finds the "Chapter" divs?

I've tried `//text//div/div`, `//text//div/p`, and `//text//div [attribute::chapter]` among many but obviously none of them are correct.

So I was wondering if anyone might provide some insight into how to find all of the "Chapter" divs?

**Comment by zme1** [comment]
Your last attempt is on the right track. You can find a `<div>` that contains a `@type` attribute with

`//div[@type]`
which is the same as

`//div[attribute::type]`
and most people use the "@" because it's shorter and easier to read.

Those paths find all `<div>` elements that have the attribute in question because they're just testing for its existence. If you want to test what it is, you need to include a comparison in your test. In an effort to point you toward the answer without giving it away, let's consider that

`div[p]`
finds all of the `<div>` elements that have any `<p>` children, while

`div[count(p) = 3]`
uses a different type of test in the predicate to finds the `<div>` elements that contain exactly three `<p>` children. Can you do something analogous to go from finding `<div>` elements that contain a `@type` attribute to those where the `@type` attribute has a particular value?

**Comment by zme1** [comment]
Thanks to David's hint, I figured out some things to do with the use of square brackets, but I want to say that I need to learn more about the use of square brackets and slashes and parentheses in constructing XPath expressions...I tend to get hung up on defining sequences when I actually don't need to...

But I'm wondering about how many ways one can go about specifying more than X amount of something. I really wanted to use the ceiling expression for number 3, and couldn't get it to work... I tried all kinds of variations with ceiling, putting it in front of the whole expression, at the end of it, etc., and finally gave up on it and did something much easier with [count].... Is there a reason I couldn't get ceiling to work--or is it my own ignorance/inexperience with XPath syntax?
I hope we'll talk about this in class tomorrow!

**Comment by zme1** [comment]
I don't think I've ever used the ceiling() function in Real Life, and I can't think of a way to use it in the current homework. Michael Kay gives an example where it might prove useful in the wild.

**Comment by zme1** [comment]
My mistake...I was fishing for terms on w3 schools, and was dazzled by their explanation: They said it "returns the smallest integer that is greater than the number argument" and somehow I didn't look at their example--to do with decimal numbers (3.14 to return the next higher or lower whole number). Yes, Michael Kay explains it clearly. So ceiling() didn't work, probably because I wasn't targeting attributes that are numbers with decimal places. I was trying to set a value for a "ceiling" at 180, as in [ceiling (180)], and applying that to limit my results only to divs that had more than 180 p elements, and of course the XPath world just laughed at me in gibberish.

**Comment by zme1** [comment]
I ended up getting a bunch of different answers for this question, but I'm still not sure if any of them are right. I know that some of my specific problems were more related to placement, which I might have figured out, but my method of finding the answer to the second question was completely different from my first combination.

for the first section, I used.

`//div[@type= "Chapter"]/p[min(180)]`
which did what I wanted in giving me the instances only once per div, but then to count them, I ended up using

`//div[@type= "Chapter"]/p[position() > 180]`
to include all of the subelements p that were over 180, within the divs, and

`count(//div[@type= "Chapter"]/p[position() > 180])`
to count them, but what I'm not sure of is what the major differences between them are, and which one of the first two gets me closer to what I want to do, which is include the divs themselves.
